### ISSUE TITLE
Is ServeDNS thread safe ?

### ISSUE THREAD

**Issue by comzyh** [shawn1m/overture#134]
In `inbound/server.go`, you are using struct field to pass parameters
https://github.com/shawn1m/overture/blob/058f993599552856261fde7094703479e9fa641c/core/inbound/server.go#L123-L127
As far as I know, `miekg/dns`'s ServeDNS is not a 'single thread' function. So,  one `ServeDNS` may be called before another `ServeDNS` finished. And, you do not use any lock there.

So, multiple flying ServeDNS may use the same dispatcher at a time. I confirmed that by logging out goroutine id at the beginning of `ServeDNS`.

Is this a problem? or Why this is not a problem?

**Comment by qcts33** [comment]
Maybe related, I'm having random crash with the following error message
I'm using 1.5rc5

```
panic: runtime error: index out of range

goroutine 2510 [running]:
github.com/miekg/dns.(*Msg).CopyTo(0xc000423680, 0xc0003626c0, 0xc000362630)
        /home/travis/gopath/src/github.com/miekg/dns/msg.go:1117 +0x49a
github.com/miekg/dns.(*Msg).Copy(...)
        /home/travis/gopath/src/github.com/miekg/dns/msg.go:1081
github.com/shawn1m/overture/core/outbound/clients.NewCacheClient(...)
        /home/travis/gopath/src/github.com/shawn1m/overture/core/outbound/clients/client_cache.go:27
github.com/shawn1m/overture/core/outbound/clients.(*RemoteClient).ExchangeFromCache(0xc00020cc80, 0x0)
        /home/travis/gopath/src/github.com/shawn1m/overture/core/outbound/clients/client_remote.go:62 +0x71
github.com/shawn1m/overture/core/outbound/clients.(*RemoteClientBundle).ExchangeFromCache(0xc0003a0600, 0x0)
        /home/travis/gopath/src/github.com/shawn1m/overture/core/outbound/clients/client_remote_bundle.go:90 +0x70
github.com/shawn1m/overture/core/outbound.(*Dispatcher).Exchange(0xc000066660, 0x5)
        /home/travis/gopath/src/github.com/shawn1m/overture/core/outbound/dispatcher.go:53 +0x222
github.com/shawn1m/overture/core/inbound.(*Server).ServeDNS(0xc000066640, 0x842660, 0xc00021b080, 0xc000429dd0)
        /home/travis/gopath/src/github.com/shawn1m/overture/core/inbound/server.go:137 +0x204
github.com/miekg/dns.(*ServeMux).ServeDNS(0xc00005b260, 0x842660, 0xc00021b080, 0xc000429dd0)
        /home/travis/gopath/src/github.com/miekg/dns/serve_mux.go:128 +0x64
github.com/miekg/dns.(*Server).serveDNS(0xc0000f2200, 0xc00021b080)
        /home/travis/gopath/src/github.com/miekg/dns/server.go:687 +0x29f
github.com/miekg/dns.(*Server).serve(0xc0000f2200, 0xc00021b080)
        /home/travis/gopath/src/github.com/miekg/dns/server.go:572 +0x2a9
github.com/miekg/dns.(*Server).worker(0xc0000f2200, 0xc00021ad00)
        /home/travis/gopath/src/github.com/miekg/dns/server.go:269 +0x1c6
created by github.com/miekg/dns.(*Server).spawnWorker
        /home/travis/gopath/src/github.com/miekg/dns/server.go:284 +0x8a
```

**Comment by shawn1m** [comment]
There is thread safe problem in dispatcher indeed and I just push the commit https://github.com/shawn1m/overture/commit/78c0acee86d0665778763557ac9111ce1ec2efd4 to quick fix it, any suggestions?

**Comment by comzyh** [comment]
just read from phone.

this quick fix is indeed an improvment.

is the assignment deep copy or not?

if not, though question and response is copied, other problems may still
exist. (or may not, it's fixed)

My suggestion is passing the message through function args all the path.

I may create an PR to fix.

Shawn <notifications@github.com> 于 2019年3月6日周三 21:09写道：

**Comment by shawn1m** [comment]
The assignment in go is just a shallow copy but most of the copied fields of dispatcher are read-only after this.

Making overture more thread safer is welcomed. Thanks a lot!

**Comment by comzyh** [comment]
I have a question, review #131, the author claim that 'GetResponseMessage' will return `nil` even if `primaryResponse` is not nil

https://github.com/shawn1m/overture/blob/ca111c9e5c0d58e671c66feaa05907a807aa0c08/core/outbound/dispatcher.go#L122-L133

But I think it's not possible. Because `RemoteClientBundle.Exchange` always return `cb.responseMessage`, the same as`GetResponseMessage`  return 
https://github.com/shawn1m/overture/blob/ca111c9e5c0d58e671c66feaa05907a807aa0c08/core/outbound/clients/client_remote_bundle.go#L85
https://github.com/shawn1m/overture/blob/ca111c9e5c0d58e671c66feaa05907a807aa0c08/core/outbound/clients/client_remote_bundle.go#L113-L115

If `primaryResponse` is not nil, `GetResponseMessage` should not return nil.

The only explanation I can make is that the author *was facing thread safe issue*, multiple DNS queries may have conflict using  PrimaryClientBundle.

Considering the thread safe issue is fixed by 78c0ace, can I remove this part of code using PR #136

**Comment by shawn1m** [comment]
It seems unnecessary and redundant now, I'll fix it sometime later.
